app/src/main/java/org/wikipedia/csrf/CsrfTokenClient.kt

```diff
@@ -0,0 +1,87 @@
+package org.wikipedia.csrf
+
+import io.reactivex.rxjava3.core.Observable
+import io.reactivex.rxjava3.schedulers.Schedulers
+import org.wikipedia.WikipediaApp
+import org.wikipedia.auth.AccountUtil
+import org.wikipedia.dataclient.Service
+import org.wikipedia.dataclient.ServiceFactory
+import org.wikipedia.dataclient.WikiSite
+import org.wikipedia.events.LoggedOutInBackgroundEvent
+import org.wikipedia.login.LoginClient
+import org.wikipedia.settings.Prefs
+import org.wikipedia.util.log.L
+import java.io.IOException
+import java.util.concurrent.Semaphore
+
+class CsrfTokenClient(private val loginWikiSite: WikiSite, private val numRetries: Int,
+                      private val csrfService: Service) {
+    constructor(site: WikiSite) : this(site, site)
+    constructor(csrfWikiSite: WikiSite, loginWikiSite: WikiSite) : this(loginWikiSite, MAX_RETRIES, ServiceFactory.get(csrfWikiSite))
+
+    val token: Observable<String>
+        get() {
+            return Observable.create { emitter ->
+                var token = ""
+                try {
+                    MUTEX.acquire()
+                    if (emitter.isDisposed) {
+                        return@create
+                    }
+                    for (retry in 0 until numRetries) {
+                        if (retry > 0) {
+                            // Log in explicitly
+                            LoginClient().loginBlocking(loginWikiSite, AccountUtil.userName!!, AccountUtil.password!!, "")
```

Comment:
Since the `loginBlocking()` requires `NonNull` username and password, would it be better if we have null checks on line `32`?

Reply:
I'm not sure how that would be helpful.  This entire function already assumes that the user is logged in, so if `userName` or `password` turns out to be null, I would want to know about it via an error passed to the emitter.

Reply:
I see, it makes sense to me.